File: packages/protocol/src/entity.ts
// entity/*, journal_entry/rescan, message/search, and recurrence/preview
// wire schemas (ADR-0009 hand-mirror).

import { Schema as S } from "effect";

/** `recurrence/preview` params (ADR-0039 amendment, #227): a draft Recurrence
 * Rule + the editing Todo's current `defer_at`/`due_at`. The editor sends this
 * read-only request to preview where the next occurrence would land. `recurrence`
 * is the opaque rule object (Core's date math validates it via a fail-safe, never
 * rejects here); the dates are optional because a Todo may carry only one anchor.
 * Hand-authored wire params (what Web sends). */
export const RecurrencePreviewParams = S.Struct({
	recurrence: S.Unknown,
	defer_at: S.optional(S.String),
	due_at: S.optional(S.String),
});

export type RecurrencePreviewParams = S.Schema.Type<
	typeof RecurrencePreviewParams
>;

/** `recurrence/preview` result (ADR-0039 amendment, #227): the next occurrence's
 * dates, or `ended: true` when completing the Todo would spawn no successor (end
 * condition reached, or a malformed/partial draft rule). `ended: true` is a normal
 * result, not an error. When `ended` is false, `defer_at`/`due_at` mirror the
 * input's anchor presence; each is omitted (not null) when absent. */
export const RecurrencePreviewResult = S.Struct({
	ended: S.Boolean,
	defer_at: S.optional(S.String),
	due_at: S.optional(S.String),
});

export type RecurrencePreviewResult = S.Schema.Type<
	typeof RecurrencePreviewResult
>;

// entity/* (ADR-0004): the accepted Entities the Library reads; `entity/list` is type-parameterized (one type per call).

/** `entity/list` params: the Entity type to list (one type per call). */
export const EntityListParams = S.Struct({ type: S.String });

export type EntityListParams = S.Schema.Type<typeof EntityListParams>;

export const ResolvedEntityRef = S.Struct({
	id: S.String,
	source_entity_id: S.String,
	target_entity_id: S.String,
	target_entity_type: S.Literal("person", "project", "todo"),
	target_title: S.optional(S.String),
	label_snapshot: S.optional(S.String),
});

export type ResolvedEntityRef = S.Schema.Type<typeof ResolvedEntityRef>;

/**
 * One Todo Person Reference on a Todo `entity/list` row (ADR-0031, ADR-0032):
 * the task-relationship analogue of `refs`. `role` carries the GTD semantics
 * (`waiting_on` ⊇ `related`). Clients derive Project↔Person↔Todo from these.
 */
export const TodoPersonRefView = S.Struct({
	person_id: S.String,
	role: S.Literal("waiting_on", "related"),
});

export type TodoPersonRefView = S.Schema.Type<typeof TodoPersonRefView>;

/**
 * One Entity's origin provenance on an `entity/list` row ("Captured from",
 * ADR-0030). A FLAT optional shape, safe because Core is the sole producer and
 * fills it from one `entity_sources` row whose CHECK guarantees exactly one
 * source kind: a user Message source carries `thread_id` + `thread_title` (link
 * back to the Thread) plus the capturing `message_id` (so the Client can
 * deep-link to the exact message, #184); a Journal-Entry source carries
 * `journal_entry_id` (link to it in the Library). Read `journal_entry_id` first,
 * else the Thread fields (`message_id` rides along with them).
 */
export const EntitySourceView = S.Struct({
	thread_id: S.optional(S.String),
	thread_title: S.optional(S.String),
	message_id: S.optional(S.String),
	journal_entry_id: S.optional(S.String),
});

export type EntitySourceView = S.Schema.Type<typeof EntitySourceView>;

/** One Entity row in an `entity/list` result: the raw tier-2 `entities` columns (ADR-0004). */
export const EntityRow = S.Struct({
	id: S.String,
	type: S.String,
	data: S.Unknown,
	created_at: S.Number,
	updated_at: S.Number,
	refs: S.optional(S.Array(ResolvedEntityRef)),
	/** Present on Todo rows: the Todo's Person References (ADR-0032). */
	person_refs: S.optional(S.Array(TodoPersonRefView)),
	/** The Entity's origin provenance (ADR-0030); absent for a user-authored Entity. */
	source: S.optional(EntitySourceView),
});

export type EntityRow = S.Schema.Type<typeof EntityRow>;

/** `entity/list` result: the accepted Entities of the requested type, newest-first. */
export const EntityListResult = S.Struct({ entities: S.Array(EntityRow) });

export type EntityListResult = S.Schema.Type<typeof EntityListResult>;

/** `entity/backlinks` params (ADR-0050): the Entity whose reverse relations the detail Inspector wants. */
export const EntityBacklinksParams = S.Struct({ entity_id: S.String });

export type EntityBacklinksParams = S.Schema.Type<typeof EntityBacklinksParams>;

/**
 * `entity/backlinks` result (ADR-0050): the two reverse sets Core resolves for the
 * detail Inspector — `mentioned_in` (distinct Journal Entries referencing this
 * Entity) and `linked_todos` (Todos linked via `project_id` / `person_refs`).
 * Both arrays are always present (possibly empty); reuses {@link EntityRow}.
 */
export const EntityBacklinksResult = S.Struct({
	mentioned_in: S.Array(EntityRow),
	linked_todos: S.Array(EntityRow),
});

export type EntityBacklinksResult = S.Schema.Type<typeof EntityBacklinksResult>;

/**
 * `entity/mutate` params (ADR-0033): a user-initiated CRUD request. `payload` is the
 * same discriminated `{mutation_kind, payload}` envelope the Worker's
 * `propose_workspace_mutation` tool uses (minus rationale), so it stays opaque at the
 * wire boundary — Core validates it per `mutation_kind`.
 */
export const EntityMutateParams = S.Struct({
	mutation_kind: S.String,
	payload: S.Unknown,
});

export type EntityMutateParams = S.Schema.Type<typeof EntityMutateParams>;

/** `entity/mutate` result: the affected Entity id — present on create/update, absent on delete. */
export const EntityMutateResult = S.Struct({
	entity_id: S.optional(S.String),
});

export type EntityMutateResult = S.Schema.Type<typeof EntityMutateResult>;

/**
 * `journal_entry/rescan` params (ADR-0042): the Journal Entry to re-scan for
 * people/projects/tasks mentioned but not yet captured. Core resolves the JE's
 * origin Thread and starts an ordinary agent Run there.
 */
export const JournalEntryRescanParams = S.Struct({
	je_id: S.String,
});

export type JournalEntryRescanParams = S.Schema.Type<
	typeof JournalEntryRescanParams
>;

/** `journal_entry/rescan` result: the spawned Run and the origin Thread it runs in (the Client follows `run/subscribe(run_id)` and can navigate to the Thread). */
export const JournalEntryRescanResult = S.Struct({
	run_id: S.String,
	thread_id: S.String,
});

export type JournalEntryRescanResult = S.Schema.Type<
	typeof JournalEntryRescanResult
>;

// message/* (ADR-0035): full-text search over completed Message text, surfaced in ⌘K.

/** One message-search hit (ADR-0035): a completed Message matching the substring query, with a SQL-rendered snippet and its Thread title for navigation. */
export const MessageHit = S.Struct({
	message_id: S.String,
	thread_id: S.String,
	run_id: S.String,
	role: S.Literal("user", "assistant"),
	snippet: S.String,
	thread_title: S.String,
	created_at: S.Number, // ms-epoch
});

export type MessageHit = S.Schema.Type<typeof MessageHit>;

/** `message/search` params (ADR-0035): a substring query over completed message text. */
export const MessageSearchParams = S.Struct({ query: S.String });

export type MessageSearchParams = S.Schema.Type<typeof MessageSearchParams>;

/** `message/search` result: matching hits, newest-first. */
export const MessageSearchResult = S.Struct({ hits: S.Array(MessageHit) });

export type MessageSearchResult = S.Schema.Type<typeof MessageSearchResult>;

File: crates/core/src/protocol/entity.rs
//! `entity/*`, `journal_entry/rescan`, and `message/search` wire types
//! (ADR-0009 hand-mirror).

use serde::{Deserialize, Serialize};

use crate::mutation::{DirectMutationKind, EntityTypeName};

/// `entity/list` params: the Entity `type` to list, one per call (e.g.
/// `"person"`). `r#type` serializes as the wire field `"type"`.
#[derive(Debug, Deserialize)]
pub struct EntityListParams {
    pub r#type: EntityTypeName,
}

/// One Entity row in `entity/list` (ADR-0004 tier-2 `entities` columns).
/// `r#type` serializes as `"type"`; `data` is the opaque entity JSON;
/// `created_at`/`updated_at` are ms-epoch stamps.
#[derive(Debug, Serialize)]
pub struct EntityRow {
    pub id: String,
    pub r#type: String,
    pub data: serde_json::Value,
    pub created_at: i64,
    pub updated_at: i64,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub refs: Vec<ResolvedEntityRef>,
    /// The Entity's origin provenance ("Captured from", ADR-0030). Omitted for a
    /// user-authored Entity (a direct Library write records no source row).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source: Option<EntitySourceView>,
}

/// One Entity's origin provenance on an `entity/list` row (ADR-0030). A FLAT
/// optional shape, safe because Core is the sole producer and fills the fields
/// from one `entity_sources` row whose CHECK guarantees exactly one source kind:
/// a user Message source carries `thread_id` + `thread_title` (link back to the
/// Thread) plus the capturing `message_id` (so the Client can deep-link to the
/// exact message, #184); a Journal-Entry source carries `journal_entry_id` (link
/// to it in the Library). The Client reads `journal_entry_id` first, else the
/// Thread fields (`message_id` rides along with them).
#[derive(Debug, Serialize)]
pub struct EntitySourceView {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thread_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thread_title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub journal_entry_id: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ResolvedEntityRef {
    pub id: String,
    pub source_entity_id: String,
    pub target_entity_id: String,
    pub target_entity_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label_snapshot: Option<String>,
}

/// `entity/list` result: the accepted Entities of the requested type,
/// newest-first. Object-wrapper shape (`{entities: [...]}`) keeps it
/// forward-extensible.
#[derive(Debug, Serialize)]
pub struct EntityListResult {
    pub entities: Vec<EntityRow>,
}

/// `entity/backlinks` params (ADR-0050): the Entity whose reverse relations the
/// detail Inspector wants. Only Person/Project are `entity_ref` targets, so
/// only those fire the read.
#[derive(Debug, Deserialize)]
pub struct EntityBacklinksParams {
    pub entity_id: String,
}

/// `entity/backlinks` result (ADR-0050): the reverse set Core resolves
/// authoritatively for the detail Inspector — `mentioned_in` (distinct Journal
/// Entries referencing this Entity, newest-occurred first). Reuses `EntityRow`
/// (ADR-0032), so the section parses through the existing entity codec. The
/// array is ALWAYS present (possibly empty `[]`); object-wrapper shape modeled
/// like `EntityListResult` for forward-extensibility.
#[derive(Debug, Serialize)]
pub struct EntityBacklinksResult {
    pub mentioned_in: Vec<EntityRow>,
}

/// `entity/mutate` params (ADR-0033): a user-initiated CRUD request. `payload` is
/// the same discriminated `{mutation_kind, payload}` envelope the Worker's
/// `propose_workspace_mutation` tool uses (minus `rationale`), so it stays opaque
/// at the wire boundary — Core validates it per `mutation_kind`.
#[derive(Debug, Deserialize)]
pub struct EntityMutateParams {
    pub mutation_kind: DirectMutationKind,
    pub payload: serde_json::Value,
}

/// `entity/mutate` result: the affected Entity id — present on create/update,
/// absent on delete (which leaves no row).
#[derive(Debug, Serialize)]
pub struct EntityMutateResult {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub entity_id: Option<String>,
}

/// `journal_entry/rescan` params (ADR-0042): the Journal Entry to re-scan for
/// people/projects/tasks mentioned but not yet captured. Core resolves the JE's
/// origin Thread and starts an ordinary agent Run there.
#[derive(Debug, Deserialize)]
pub struct JournalEntryRescanParams {
    pub je_id: String,
}

/// `journal_entry/rescan` result: the spawned Run and the origin Thread it runs
/// in (so the Client can follow `run/subscribe(run_id)` and navigate to the
/// Thread). Mirror of TS `JournalEntryRescanResult`.
#[derive(Debug, Serialize)]
pub struct JournalEntryRescanResult {
    pub run_id: String,
    pub thread_id: String,
}

/// `message/search` params (ADR-0035): a substring query over completed Message
/// text. Mirror of TS `MessageSearchParams`.
#[derive(Debug, Deserialize)]
pub struct MessageSearchParams {
    pub query: String,
}

/// One `message/search` hit (ADR-0035): a completed Message matching the
/// substring query, with a SQL-rendered snippet and its Thread title for
/// navigation. Mirror of TS `MessageHit` (field-for-field, snake_case wire);
/// aligns with `db::MessageHit`. `role` is `"user"`/`"assistant"` on the wire;
/// `created_at` is a ms-epoch stamp.
#[derive(Debug, Serialize)]
pub struct MessageHit {
    pub message_id: String,
    pub thread_id: String,
    pub run_id: String,
    pub role: String,
    pub snippet: String,
    pub thread_title: String,
    pub created_at: i64,
}

/// `message/search` result: the matching hits, newest-first. Object-wrapper
/// shape (`{hits: [...]}`) keeps it forward-extensible. Mirror of TS
/// `MessageSearchResult`.
#[derive(Debug, Serialize)]
pub struct MessageSearchResult {
    pub hits: Vec<MessageHit>,
}

/// Mirror tests: lock the Rust serde shapes to the canonical snake_case wire
/// JSON the TS `Schema` definitions in `packages/protocol` produce (ADR-0009).
/// Each test asserts agreement in the type's available direction; a renamed
/// field or changed type fails the matching test. This is the reconciliation
/// point that guards against TS/Rust divergence.
#[cfg(test)]
mod mirror_tests {
    use super::*;
    use serde_json::json;

    // A fixed UUID-shaped string; the wire carries ids as plain strings.
    const UUID_A: &str = "0190d3c1-0000-7000-8000-000000000001";
    const UUID_B: &str = "0190d3c1-0000-7000-8000-000000000002";
    const UUID_RUN: &str = "0190d3c1-0000-7000-8000-000000000003";

    #[test]
    fn entity_params_reject_values_outside_their_closed_domains() {
        assert!(serde_json::from_value::<EntityListParams>(json!({ "type": "todo" })).is_err());
        assert!(
            serde_json::from_value::<EntityMutateParams>(json!({
                "mutation_kind": "create_todo",
                "payload": {}
            }))
            .is_err()
        );
        assert!(
            serde_json::from_value::<EntityMutateParams>(json!({
                "mutation_kind": "apply_intent_graph",
                "payload": {}
            }))
            .is_err()
        );
    }

    #[test]
    fn message_search_params_rejects_missing_and_non_string_query() {
        assert!(serde_json::from_value::<MessageSearchParams>(json!({})).is_err());
        assert!(serde_json::from_value::<MessageSearchParams>(json!({ "query": 42 })).is_err());
    }

    #[test]
    fn message_search_result_encodes_hits_wrapper_and_empty() {
        let one = MessageSearchResult {
            hits: vec![MessageHit {
                message_id: UUID_A.to_string(),
                thread_id: UUID_B.to_string(),
                run_id: UUID_RUN.to_string(),
                role: "user".to_string(),
                snippet: "hi".to_string(),
                thread_title: "T".to_string(),
                created_at: 1,
            }],
        };
        let value = serde_json::to_value(&one).unwrap();
        assert_eq!(value["hits"].as_array().unwrap().len(), 1);
        assert_eq!(value["hits"][0]["role"], json!("user"));

        let empty = MessageSearchResult { hits: vec![] };
        assert_eq!(serde_json::to_value(&empty).unwrap(), json!({ "hits": [] }));
    }
}
